Documentation at docs.flutter.dev/go/android-java-gradle-error is incorrect when using newer versions of Java

For example, trying to run flutter with Java 20 and any gradle version less than 8.1 will result in the following error:
```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':gradle:compileGroovy'.
> BUG! exception in phase 'semantic analysis' in source unit '/Users/mackall/development/flutter/flutter/packages/flutter_tools/gradle/src/main/groovy/flutter.groovy' Unsupported class file major version 64

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 15s
Running Gradle task 'assembleDebug'...                             16.4s

┌─ Flutter Fix ─────────────────────────────────────────────────────────────────────────────────────────────┐
│ [!] Your project's Gradle version is incompatible with the Java version that Flutter is using for Gradle. │
│                                                                                                           │
│ To fix this issue, consult the migration guide at docs.flutter.dev/go/android-java-gradle-error.          │
└───────────────────────────────────────────────────────────────────────────────────────────────────────────┘
Exception: Gradle task assembleDebug failed with exit code 1
```

The documentation at the above link in many places specifically calls out upgrading to Gradle 7.6.1, but this will not fix the issue if you are using Java 20 (and also would probably be a downgrade in this case). 

We should either avoid providing this error message when using a newer Java version than 17, or make sure that the documentation is helpful for all Java versions (the way to make it helpful would be to tell users to consult the Gradle [compatibility matrix](https://docs.gradle.org/current/userguide/compatibility.html) and upgrade by picking a Gradle version that fits their java version).

--- a/packages/flutter_tools/lib/src/android/gradle_errors.dart
+++ b/packages/flutter_tools/lib/src/android/gradle_errors.dart
@@ -685,12 +685,24 @@ final GradleHandledError incompatibleJavaAndGradleVersionsHandler = GradleHandle
     required bool usesAndroidX,
     required bool multidexEnabled,
   }) async {
+    final File gradlePropertiesFile = project.directory
+        .childDirectory('android')
+        .childDirectory('gradle')
+        .childDirectory('wrapper')
+        .childFile('gradle-wrapper.properties');
     // TODO(reidbaker): Replace URL with constant defined in
     // https://github.com/flutter/flutter/pull/123916.
     globals.printBox(
       "${globals.logger.terminal.warningMark} Your project's Gradle version "
       'is incompatible with the Java version that Flutter is using for Gradle.\n\n'
-      'To fix this issue, consult the migration guide at docs.flutter.dev/go/android-java-gradle-error.',
+          'If you recently upgraded Android Studio, consult the migration guide '
+          'at docs.flutter.dev/go/android-java-gradle-error.\n\n'
+          'Otherwise, to fix this issue, first, check the Java version used by Flutter by '
+          'running `flutter doctor --verbose`.\n\n'
+          'Then, update the Gradle version specified in ${gradlePropertiesFile.path} '
+          'to be compatible with that Java version. '
+          'See the link below for more information on compatible Java/Gradle versions:\n'
+          'https://docs.gradle.org/current/userguide/compatibility.html#java\n\n',
       title: _boxTitle,
     );
     return GradleBuildStatus.exit;

--- a/packages/flutter_tools/lib/src/android/gradle_errors.dart
+++ b/packages/flutter_tools/lib/src/android/gradle_errors.dart
@@ -694,7 +694,7 @@ final GradleHandledError incompatibleJavaAndGradleVersionsHandler = GradleHandle
     // https://github.com/flutter/flutter/pull/123916.
     globals.printBox(
       "${globals.logger.terminal.warningMark} Your project's Gradle version "
-      'is incompatible with the Java version that Flutter is using for Gradle.\n\n'
+          'is incompatible with the Java version that Flutter is using for Gradle.\n\n'
           'If you recently upgraded Android Studio, consult the migration guide '
           'at docs.flutter.dev/go/android-java-gradle-error.\n\n'
           'Otherwise, to fix this issue, first, check the Java version used by Flutter by '

--- a/packages/flutter_tools/test/general.shard/android/gradle_errors_test.dart
+++ b/packages/flutter_tools/test/general.shard/android/gradle_errors_test.dart
@@ -1376,9 +1376,12 @@ Could not compile build file '…/example/android/build.gradle'.
         multidexEnabled: true,
       );
 
-      // Ensure the error notes the incompatible Gradle/AGP/Java versions and links to related resources.
+      // Ensure the error notes the incompatible Gradle/AGP/Java versions, links to related resources,
+      // and a portion of the path to where to change their gradle version.
       expect(testLogger.statusText, contains('Gradle version is incompatible with the Java version'));
       expect(testLogger.statusText, contains('docs.flutter.dev/go/android-java-gradle-error'));
+      expect(testLogger.statusText, contains('gradle-wrapper.properties'));
+      expect(testLogger.statusText, contains('https://docs.gradle.org/current/userguide/compatibility.html#java'));
     }, overrides: <Type, Generator>{
       GradleUtils: () => FakeGradleUtils(),
       Platform: () => fakePlatform('android'),

